packages/react-checkbox/src/components/Checkbox/useCheckbox.tsx

```diff
@@ -1,101 +1,94 @@
-export const useCheckbox = (props: CheckboxProps, ref: React.Ref<HTMLElement>): CheckboxState => {
-  const { disabled = false, circular = false, required = false, id, rootId, onChange: userOnChange } = props;
+export const useCheckbox = (props: CheckboxProps, ref: React.Ref<HTMLInputElement>): CheckboxState => {
+  const { children, circular, size = 'medium', labelPosition = 'after' } = props;
+
   const [checked, setChecked] = useControllableState({
     defaultState: props.defaultChecked,
     state: props.checked,
     initialState: false,
   });
 
-  const inputInternalRef = React.useRef<HTMLInputElement>(null);
-  const inputShorthand = resolveShorthand(props.input, {
-    required: true,
-    defaultProps: {
-      disabled,
-      type: 'checkbox',
-      required,
-    },
-  });
-
   const state: CheckboxState = {
+    children,
     circular,
     checked,
-    size: 'medium',
-    labelPosition: 'after',
-    rootId,
+    size,
+    labelPosition,
     components: {
       root: props.children !== undefined ? (Label as React.ComponentType<LabelProps>) : 'span',
       indicator: 'div',
       input: 'input',
     },
-    input: {
-      ...inputShorthand,
-      ref: useMergedRefs(inputShorthand.ref, inputInternalRef),
-    },
+    root: resolveShorthand(props.root, {
+      required: true,
+      defaultProps: getRootSlotNativeProps(props),
+    }),
+    input: resolveShorthand(props.input, {
+      required: true,
+      defaultProps: {
+        type: 'checkbox',
+        id: useId('checkbox-', props.id),
+        ref,
+        // Exclude checked/defaultChecked/children from the input slot; they have custom handling in useCheckbox
+        ...getPrimarySlotNativeProps('input', props, /* excludedPropNames:*/ ['checked', 'defaultChecked', 'children']),
+      },
+    }),
     indicator: resolveShorthand(props.indicator, {
       required: true,
     }),
-    root: getNativeElementProps('div', { ref, ...props }),
   };
 
-  state.input.checked = checked === true;
-  state.checked = checked ? checked : false;
-
-  if (state.indicator && !state.indicator.children) {
+  if (!state.indicator.children) {
     if (state.size === 'medium') {
       state.indicator.children = checked === 'mixed' ? <Mixed12Regular /> : <Checkmark12Regular />;
     } else {
       state.indicator.children = checked === 'mixed' ? <Mixed16Regular /> : <Checkmark16Regular />;
     }
   }
 
-  const inputOnChange = state.input.onChange;
+  const onChange = state.input.onChange as CheckboxProps['onChange'];
   state.input.onChange = useEventCallback(ev => {
     ev.stopPropagation();
-    inputOnChange?.(ev);
-
     const val = ev.currentTarget.indeterminate ? 'mixed' : ev.currentTarget.checked;
-
-    userOnChange?.(ev, { checked: val });
+    onChange?.(ev, { checked: val });
     setChecked(val);
   });
 
-  state.input.id = useId('checkbox-', id);
-  state.root.id = state.rootId;
+  const inputRef = useMergedRefs(state.input.ref);
+  state.input.ref = inputRef;
```

Comment:
`useMergedRefs` returns a new ref-like object with a `.current` prop, which is all I need here. It seems unnecessary to create another ref as well?

I can't use `state.input.ref` directly because it might be unset or a function ref.

Reply:
Observation: @behowell is right, it makes usage of `state.input.ref` safe, but I also agree with @bsunderhus that it becomes unclear. I think that it's reasonable to add a hook named `useSafeRef()` (_other suggestions?_) that will solve this problem without merging refs. WDYT?

_Of course, it's out of scope of this PR._

Reply:
Calling it a "merged" ref makes more sense than a "safe" ref to me (perhaps singular `useMergedRef` would make slightly more sense). It creates a new ref object that is merged with the ref I passed in. I can then use that merged ref for whatever I need.

Reply:
If we're going to use it as a way of normalizing refs than I'm not sure `useMergedRefs` is the proper name also, but yeah, @layershifter  is right, it's out of this scope anyway.

IMO the `useMergedRefs` name implies you're merging multiple refs that comes as arguments into a single one, that's more of the react way of doing things I guess... on the other hand the way we're using right now seems like a lot of mutations are happening.